zbench.zig

```diff
@@ -183,117 +244,88 @@ pub const Benchmark = struct {
-            func(bench);
-        }
+pub const Percentiles = struct {
+    p75: u64,
+    p99: u64,
+    p995: u64,
+};
 
-        bench.stop();
-        // double N for next iteration
-        if (bench.N < MAX_N / 2) {
-            bench.N *= 2;
-        } else {
-            bench.N = MAX_N;
-        }
+/// BenchmarkResult stores the resulting computed metrics/statistics from a benchmark
+pub const BenchmarkResult = struct {
+    name: []const u8,
+    percs: Percentiles,
+    avg_duration: usize,
+    std_duration: usize,
+    min_duration: usize,
+    max_duration: usize,
+    total_operations: usize,
+
+    pub fn prettyPrint(self: BenchmarkResult, header: bool) !void {
+        var p75_buffer: [128]u8 = undefined;
+        const p75_str = try format.duration(p75_buffer[0..], self.percs.p75);
 
-        iterations += 1; // Increase the iteration counter
-        duration += bench.elapsed(); // ...and duration
-    }
+        var p99_buffer: [128]u8 = undefined;
+        const p99_str = try format.duration(p99_buffer[0..], self.percs.p99);
 
-    // Safety first: make sure the recorded durations aren't all-zero
-    if (duration == 0) duration = 1;
-
-    // Adjust N based on the actual duration achieved
-    bench.N = @intCast((bench.N * MIN_DURATION) / duration);
-    // check that N doesn't go out of bounds
-    if (bench.N == 0) bench.N = 1;
-    if (bench.N > MAX_N) bench.N = MAX_N;
-
-    // Now run the benchmark with the adjusted N value
-    bench.reset();
-    var j: usize = 0;
-    while (j < bench.N) : (j += 1) {
-        bench.start();
-        func(bench);
-        bench.stop();
+        var p995_buffer: [128]u8 = undefined;
+        const p995_str = try format.duration(p995_buffer[0..], self.percs.p995);
+
+        var avg_std_buffer: [128]u8 = undefined;
+        var avg_std_offset = (try format.duration(avg_std_buffer[0..], self.avg_duration)).len;
+        avg_std_offset += (try std.fmt.bufPrint(avg_std_buffer[avg_std_offset..], " ± ", .{})).len;
+        avg_std_offset += (try format.duration(avg_std_buffer[avg_std_offset..], self.std_duration)).len;
+
+        // FIXME: This looks quite ugly, maybe there's a better way to do all this?
+        var min_max_buffer: [128]u8 = undefined;
+        min_max_buffer[0] = '(';
+        var min_max_offset = (try format.duration(min_max_buffer[1..], self.min_duration)).len + 1;
+        min_max_offset += (try std.fmt.bufPrint(min_max_buffer[min_max_offset..], " ... ", .{})).len;
+        min_max_offset += (try format.duration(min_max_buffer[min_max_offset..], self.max_duration)).len;
+        min_max_buffer[min_max_offset] = ')';
+        min_max_offset += 1;
+
+        if (header) prettyPrintHeader();
+        std.debug.print("{s:<25} \x1b[90m{d:<8} \x1b[33m{s:<22} \x1b[94m{s:<28} \x1b[90m{s:<10} \x1b[90m{s:<10} \x1b[90m{s:<10}\x1b[0m\n", .{ self.name, self.total_operations, avg_std_buffer[0..avg_std_offset], min_max_buffer[0..min_max_offset], p75_str, p99_str, p995_str });
```

Comment:
std.debug.print can get removed from optimized builds, no? might want to use a stdout printer here.

Reply:
I'm not sure, I didn't even consider it. Yeah we for sure want to print to stdout, not use the debugging functionality :sweat_smile: